feat: Add check command that compares GitHub contributors with credited ones

machour opened the pull request:
More than often, I find myself having a diff between the numbers of contributors reported by Github, and the contributors actually credited in my `.all-contributorsrc`.

This PR draft introduce a new CLI command that will fetch contributors from GH api, and run a two-ways diff with the known ones, allowing a project maintainer to easily find who is missing.

Sample output on this repo:

```
🍺  ~/all-contributors-cli $ ./cli.js check jfmengels/all-contributors-cli
Missing contributors in .all-contributorsrc:
    brycereynolds, chrisinajar, jmeas, spirosikmd, fadc80, snipe
Unknown contributors found in .all-contributorsrc:
    charlike
```

I still need to paginate the api response if there are more than 100 results, and avoid reporting unknown contributors if their contributions didn't involve coding (video, blog, funding, ..)

I'm also thinking about adding a `repository` field in the configuration file, to avoid retyping the repo name every time the command is executed.

And I'm not sure if `utils/diff.js` is the right place for this script 🤔

kentcdodds commented:
This is super awesome!

> avoid reporting unknown contributors if their contributions didn't involve coding (video, blog, funding, ..)

:+1:

> I'm also thinking about adding a repository field in the configuration file, to avoid retyping the repo name every time the command is executed.

You can get that from the `projectOwner` and `projectName` fields :+1:

kentcdodds commented on cli.js:
```diff
@@ -23,6 +23,8 @@ var argv = yargs
   .usage('Usage: $0 add <username> <contribution>')
   .command('init', 'Prepare the project to be used with this tool')
   .usage('Usage: $0 init')
+  .command('check', 'Compares contributors from Github with the ones credited in .all-contributorsrc')
```
`Github` should be `GitHub` :+1:

kentcdodds commented on cli.js:
```diff
@@ -23,6 +23,8 @@ var argv = yargs
   .command('init', 'Prepare the project to be used with this tool')
   .usage('Usage: $0 init')
+  .command('check', 'Compares contributors from Github with the ones credited in .all-contributorsrc')
+  .usage('Usage: $0 check <repository>')
```
Don't forget to remove `<repository>` when you refactor this to use `projectOwner` and `projectName`

kentcdodds commented on cli.js:
```diff
@@ -87,6 +108,9 @@ function promptForCommand(argv) {
       name: 'Re-generate the contributors list',
       value: 'generate'
+    }, {
+      name: 'Compare contributors from Github with the credited ones',
```
`Github` should be `GitHub`

machour commented:
Hi @kentcdodds, thank you for the feedback. I took care of it!

I also implemented API pagination, but it's currently [limited to 500 contributors max](https://developer.github.com/v3/repos/#list-contributors). I guess that will be more than enough for most projects 😄 

For collaborators unknown to GitHub, I added a check to only report those whose contributions includes "code" or "tool". There might be room for improvement here, but I think it's enough for now.

Added some documentation too, you may want to review it as I'm not a native speaker.

And last but not least, I'm not sure how to write tests for this new feature. Aside from testing headers parsing in `getNextLink`, not really sure what I can do. I would love to have your input on that.

kentcdodds commented:
> For collaborators unknown to GitHub, I added a check to only report those whose contributions includes "code" or "tool". There might be room for improvement here, but I think it's enough for now.

Actually, let's make it `code` or `test` because `tool` is often referencing another related project and may or may not be something that happened in the repo. What do you think?

kentcdodds commented:
Just realized that the `yarn.lock` file is already in the project. That's fine. This isn't really set up the way my projects are normally. It's @jfmengels project after all :wink:

It's also using AVA rather than Jest, so I'm not certain of the best way to test things either, but I think you'll want to mock the `request` library somehow when testing the check functionality. I'm afraid I don't have too much time to give you feedback or ideas on this point :-/

machour commented:
@kentcdodds took in account your remarks and added tests using `nock()` which was already included in the project. 100% coverage for `check.js`, so if the test suite looks good for you I think I'm done with this PR: 
   https://github.com/jfmengels/all-contributors-cli/pull/58/files#diff-4c2289f0a9b1e98a126e9406cae8c04b

fixtures are in `fixtures/` which is ignored in `package.json`, the test file uses the `.test.js` suffix in its name.. I think I won't mess a thing by publishing this. (I'm still a bit new to publishing, triple checking with you 😊 )

kentcdodds commented on cli.js:
```diff
@@ -68,6 +71,37 @@ function addContribution(argv) {
+
+    var missingInConfig = ghContributors.filter(login => knownContributors.indexOf(login) === -1);
+    var missingFromGithub = knownContributors.filter(login => {
+      return ghContributors.indexOf(login) === -1 && (
```
Could probably use `includes` here :)

kentcdodds commented on cli.js:
```diff
@@ -68,6 +71,37 @@ function addContribution(argv) {
+    }, {});
+    var knownContributors = configData.contributors.map(contributor => contributor.login);
+
+    var missingInConfig = ghContributors.filter(login => knownContributors.indexOf(login) === -1);
```
Could probably use `includes` here.

kentcdodds commented on lib/util/check.test.js:
```diff
@@ -0,0 +1,48 @@
+});
+
+test('Handle a single results page correctly', async t => {
+  await check('jfmengels', 'all-contributors-cli').then(transformed => {
```
Because you're using `await` here, you don't need the `then` you could do:

```javascript
const transformed = await check ...
t.deepEqual(transformed ....
```

kentcdodds commented on lib/util/check.js:
```diff
@@ -0,0 +1,45 @@
+    return null;
+  }
+
+  var nextLink = link.split(',').find(s => s.indexOf('rel="next"') > -1);
```
Could use `includes` here

kentcdodds commented on lib/util/check.js:
```diff
@@ -0,0 +1,45 @@
+}
+
+module.exports = function getContributorsFromGithub(owner, name) {
+  var url = 'https://api.github.com/repos/' + owner + '/' + name + '/contributors?per_page=100';
```
Could use a template literal:

```js
var url = `https://api.github.com/repos/${owner}/${name}/contributors?per_page=100`;
```

machour replied:
There you go. Also use template literals for the command output

kentcdodds commented:
Great work @machour :clap:
